### ISSUE TITLE
Add a way to define functions that always run without the GIL

### ISSUE THREAD

**Issue by pablogsal** [cython/cython#3443]
When working with C and C++ extensions, is common to have functions that must run without the GIL held (for instance, to make sure there are no deadlocks that involve other locks or native threads). This can be easily achieved doing:

```
cdef void my_func_wrapper(...):
   with nogil:
          my_func(...)
```

The problem is that when exposing a large library or multiple methods and classes that fall in this situation, you end with tons of tons of wrappers like the one before.

Cython already offers a solution for the opposite problem (taking the GIL always when the function is called) using the `with gil` declaration:

```
cdef void my_func_wrapper(...) with gil
```

but there is no obvious way that I know to achieve the contrary (the original case I am discussing to drop always the GIL).

For this reason, I propose to add an analogous `with nogil` qualifier to function declarations:

```
cdef void my_func_wrapper(...) with nogil
```

that basically generates the following code:

```
static double __pyx_f_3fib_my_func(void) {
  #ifdef WITH_THREAD
  PyThreadState *_save;
  Py_UNBLOCK_THREADS
  __Pyx_FastGIL_Remember();
  #endif

// FUNCTION CODE HERE //

  /* function exit code */
  #ifdef WITH_THREAD
  __Pyx_FastGIL_Forget();
  Py_BLOCK_THREADS
  #endif
}

``

I have a draft PR  if everyone agrees that this is a desired feature to add.

**Comment by McSinyx** [comment]
`nevergil` sounds great!

**Comment by scoder** [comment]
I'm ok with adding this, but compared to the with-gil function modifier, the latter saves both a modifier (nogil) and an indented block (with gil), whereas your proposal really only seems to save one indentation (with nogil).

Actually, the with-gil modifier was there before the context manager, simply because it was much easier to implement at the time. These days, we'd probably need a bit of a discussion to add it as a feature, if it wasn't there already.

IIUC, however, your proposal also allows calling the function directly regardless of whether the GIL is held or not, right? That makes it sound appealing to me, and supports a real use case for it to be a modifier.

**Comment by McSinyx** [comment]
As I understand, the following symmetric qualifier system is proposed:
* `gil`: OK to be called with GIL (default, but such keyword is not allowed)
* `nogil`: OK to be called without GIL (already exist)
* `with gil`: must be called with GIL (by acquiring it) (also already exist)
* `with nogil`: must be called without GIL (by releasing it) (to be implemented)

With version 3.0 coming in a *can't seem to let you go, can't seem to hold you close* future, I suggest a discussion on those keywords, which does not very welly reflect the actual meaning of the modifier.

**Comment by pablogsal** [comment]
> IIUC, however, your proposal also allows calling the function directly regardless of whether the GIL is held or not, right? That makes it sound appealing to me, and supports a real use case for it to be a modifier.

Yeah, in my original idea calling the function with the gil held will release it, call the actual function and re-acquire it. On the other hand in my original idea calling this function without the GIL will just call the function without doing anything with the GIL but we can discuss what should happen if you prefer something different.

Also, one of the key ideas that I would like to explore with this is that you don't need to create the wrapper if you expose tons of functions that must be called without the GIL. For instance, this definition in a `pxd` file:

```
cdef extern from "lunch.h":
    void my_func(...) with nogil
```

will make sure that that `my_func` is always called without the GIL held, without the need to create the wrapper manually. What are your thoughts on this?

**Comment by godlygeek** [comment]
Given a function that must be called with the GIL released, something like:

```c
int email_sre_team(const char *msg);
```

If I didn't need to release the GIL, I'd be able to do:

```py
def report_server_status(str host, str status):
    if email_sre_team("Host %s is now %s" % (host, status)):
        raise RuntimeError("Couldn't send email to SREs")
```


But, because I need to release the GIL before making this network call, and because automatic type conversions can't be done while the GIL is released, that needs to be:

```py
def report_server_status(str host, str status):
    msg = "Host %s is now %s" % (host, status)
    cdef const char *c_msg = msg
    cdef int rc
    with nogil:
        rc = email_sre_team(c_msg)
    if rc:
        raise RuntimeError("Couldn't send email to SREs")
```

It would be tremendously useful if declaring a function in a `.pxd` file as `with nogil` (or whatever qualifier we settle on) would automatically drop the GIL after applying type conversions on the function's parameters and before calling the function, then reacquire it after the function call completes.

**Comment by scoder** [comment]
@godlygeek I don't think your example can benefit from any automation. There will always be tiny things to prepare and tear down around nogil blocks, and those will change over time. Making it a Cython feature to detect those and exclude them from the nogil block isn't going to work well, not for the compiler, and also not for users who have to understand the limitations.

**Comment by scoder** [comment]
@McSinyx the declarations are more like this:

-   'gil': GIL is assumed to be held, normal Python function behaviour, thus no special keyword
-   `nogil`: assumes the GIL not be be held on entry (although it may be) and prevents the function body from using operations that require the GIL (except for operations that explicitly or implicitly acquire the GIL). Requires a C function signature.
-   `with gil`: assures that the GIL is held during the function body, regardless of whether it was released before or not. Same as the `with gil:` context manager around the complete function body together with a `nogil` function declaration. Requires a C function signature.
-   `with nogil`: assures that the GIL is released around the function body, regardless of whether it was held before or not. Same as the `with nogil:` context manager around the complete function body. When used on `def` and `cpdef` functions, the Python function wrapper will always require the GIL to be held on entry. `cdef` functions and the C entry point of `cpdef` functions will automatically be marked as `nogil` and can thus be called directly from `nogil` code.

The last bit is something that may require a bit more work on top of Pablo's current WIP branch.

**Comment by scoder** [comment]
@pablogsal 
```cython
cdef extern from "lunch.h":
    void my_func(...) with nogil
```
I don't see how this should work. From the point of view of a caller, `with gil` is a marker that changes the way the function works _internally_. All it really sees from that is the implicit `nogil` declaration, i.e. it knows that it can call the function without holding the GIL. Whether that function then acquires the GIL or not is irrelevant to the caller.

The same applies to `with nogil`. Here, again, the caller only sees that the C entry point to the function is declared as `nogil` and can thus be called without holding the GIL. The Python entry point of a `cpdef` function will always require the GIL and thus doesn't change. Whether that function really releases the GIL internally or not doesn't matter.

Thus, the above declarations is really the same as `void my_func(...) nogil`, meaning that we should not support `with nogil` in extern declarations in order to avoid confusion.

**Comment by scoder** [comment]
@pablogsal @godlygeek I first wrote something different here, please ignore that.

I understand your use case regarding `extern` declarations of `with nogil` functions. Cython usually knows internally whether it really needs to free the GIL for the call (or if it's freed already), and thus what code it needs to generate around such a call or not.

One drawback is that a `nogil` function actually does *not* know whether the GIL is held or not when it's being called, and thus would _not_ release the GIL around such a call, even if the outer caller is holding the GIL.

And secondly, note that a declaration is actually a pretty broad and global thing. The fact that a function can (or should) be called without holding the GIL does not mean that all usages will always _want_ it to be called with the GIL released. Once an external function is declared as `with nogil`, as in your example, all usages would then automatically (and implicitly) try to release the GIL for it. In order to hold the GIL around the call, a user would have to redeclare the function separately (maybe even under a different name). I'm not so happy about the implicit note in this. It's not visible from the code at all in such a case whether the GIL is being held or not, which makes the concurrency implications unclear.

**Comment by scoder** [comment]
@godlygeek your example can be rewritten like this, which makes it at least a bit less verbose:
```cython
def report_server_status(str host, str status):
    msg = f"Host {host} is now {status}".encode("ascii")
    cdef const char* c_msg = msg
    with nogil:
        if email_sre_team(c_msg):
            raise RuntimeError("Couldn't send email to SREs")
```
A `raise` statement will implicitly acquire the GIL if needed, so you can use it inside of `nogil` sections.

I see an interest in the argument conversion part, but my guess is that your example is a bit of an edge case still. Other `nogil` use cases may require more than one external call, in which case you'd _want_ to release the GIL only once, not for each call. And not all conversions are so trivial that they can be written in place, e.g. the message string conversion in your example, which Cython would actually flag as illegal because you are taking a C pointer to a temporary Python object, the lifetime of which does not cover the function call into which you want to pass it.

In short, the feature may seem appealing at first sight, but I'm not sure it's worth the added complexity for real world applications.

**Comment by godlygeek** [comment]
> A `raise` statement will implicitly acquire the GIL if needed, so you can use it inside of `nogil` sections.

I didn't realize that; good to know.

> I see an interest in the argument conversion part, but my guess is that your example is a bit of an edge case still.

Perhaps.  I'm currently wrapping a library where the GIL must always be released when calling into it.  This is a bit of a simplification, but it's something like: logs from the wrapped C++ library are enqueued to a buffer, and when that buffer fills they are flushed by a background thread to the Python `logging` module.  The GIL must never be held while calling into this library, as a deadlock can occur if a log statement in the library results in a log buffer filling up and needing to be flushed: the log flushing thread needs to acquire the GIL before it can flush those logs to Python, and the foreground thread can't progress to releasing the GIL until it has enqueued the new log message.  It's not quite this simple, but that's a close enough analogy to discuss the use case.

> Other `nogil` use cases may require more than one external call, in which case you'd _want_ to release the GIL only once, not for each call.

True, but this feature wouldn't preclude you from doing that, it would just simplify things if you choose not to do that, at some performance cost.

> the message string conversion in your example, which Cython would actually flag as illegal because you are taking a C pointer to a temporary Python object

I tested first; no such warning occurred.  The generated code is:

```c
__pyx_t_2 = __Pyx_PyString_Format(__pyx_kp_s_Host_s_is_now_s, __pyx_t_1); if (unlikely(!__pyx_t_2)) __PYX_ERR(0, 390, __pyx_L1_error)
__Pyx_GOTREF(__pyx_t_2);
__Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
__pyx_t_3 = __Pyx_PyObject_AsString(__pyx_t_2); if (unlikely((!__pyx_t_3) && PyErr_Occurred())) __PYX_ERR(0, 390, __pyx_L1_error)
__pyx_t_4 = (email_sre_team(__pyx_t_3) != 0);
```

The temporary Python string at `__pyx_t_2` does stay in scope through the call to `email_sre_team`.

> `with nogil`: assures that the GIL is released around the function body, regardless of whether it was held before or not

The example I gave can actually be simplified by the use of a wrapper function with a `nogil` wrapped around the function's body:

```py
cdef int _email_sre_team_with_nogil(const char *msg):
    with nogil:
        return email_sre_team(msg)

def report_server_status(str host, str status):
    if _email_sre_team_with_nogil("Host %s is now %s" % (host, status)):
        raise RuntimeError("Couldn't send email to SREs")
```

If the proposed `with nogil` modifier can be attached to external declarations and behaves exactly as though `with nogil:` is wrapped around the body of the call, then essentially it would just be generating exactly that wrapper automatically.

**Comment by scoder** [comment]
>> Other nogil use cases may require more than one external call, in which case you'd want to release the GIL only once, not for each call.

> True, but this feature wouldn't preclude you from doing that, it would just simplify things if you choose not to do that, at some performance cost.

That's what I meant with my "declarations are global" comment above, though. Once you've declared a function as `with nogil`, it behaves that way. That limits the reuse of such declarations. For example, you couldn't publish .pxd files with such declarations on PyPI, simply because you can't know how people would use them. You're assigning a behaviour to an external function here that may or may not be desired. And this need may even differ between different usages inside of one package or module. In short, it's generally the _usage_ that decides about this kind of behavioural difference, not the _declaration_.

I'm not saying that there is no use case for this within a given limited scope, it just adds limitations that users will have to be aware of, and that may not be obvious from their current work focus.

**Comment by scoder** [comment]
> I tested first; no such warning occurred.

Ah, right. :) I keep forgetting about the stuff that I fix over time …

**Comment by godlygeek** [comment]
I think we're talking about two separate things here.

> Once you've declared a function as `with nogil`, it behaves that way. That limits the reuse of such declarations.

If I understand correctly, you're pointing out that if an external function is declared using the `with nogil` modifier, it will never be possible to continue to hold the GIL throughout a call to that external function, and that for many functions, there may be times when users want to hold the GIL while calling them.  I concede that this is the case, and that it does limit the reusability of these declarations, but I believe that there are many functions - nearly anything that does network IO, where wait times can be unbounded, for instance - where the GIL ought to be dropped on every call.  Even if it's not useful for most functions, it is useful for some.

> Other `nogil` use cases may require more than one external call, in which case you'd _want_ to release the GIL only once, not for each call.

As far as I understand, this is an entirely separate concern from the above.  Here, you're saying that, if that you want to make 3 calls in a row without the GIL, you wouldn't want to release and re-acquire the GIL 3 times, as it's far more efficient to only do it once.  That's certainly true, but the way I'm hoping that the `with nogil` modifier would work, nothing would stop the user from combining it with the `with nogil` context manager: if those 3 calls to `with nogil`-modified functions are made inside the body of the context manager, each of those 3 calls would perform the `PyEval_SaveThread()` /  `PyEval_RestoreThread` dance to drop the GIL, but since it has already been dropped by the context manager those calls are very cheap as there's no need for any synchronization across threads.

**Comment by scoder** [comment]
> Even if it's not useful for most functions, it is useful for some.

… and we've always given users the ability to shoot themselves in the foot, _if_ they want to.

I can see this being useful at least in some cases, so, why not. Let's make this a feature.

**Comment by pablogsal** [comment]
> I can see this being useful at least in some cases, so, why not. Let's make this a feature.

Should I start the support for such change in the draft PR?

**Comment by scoder** [comment]
> Should I start the support for such change in the draft PR?

My guess is that the `extern` function support isn't difficult to add on top of the general feature, and should also stay a somewhat local code change (possibly even just a new `GilStatNode` injected somewhere in `SimpleCallNode.analyse_types()`). So, I wouldn't mind. If you notice that it becomes visibly more complex than that, we should consider extracting it into a separate PR, but let's start by keeping it together for now.

**Comment by ronf** [comment]
+1 from me on this feature!

In terms of naming, what do you think of having keywords of "acquiregil" as an alternative to the existing "with gil" and "releasegil" instead of the proposed "with nogil" for the case where you want to release the GIL if it is being held before running the function and re-acquire it before returning the result?

**Comment by scoder** [comment]
@ronf Since the semantics match the existing `with gil` and `with nogil` features, I think we should not add new names.

**Comment by ronf** [comment]
Ok - that works for me. Having played with this a bit more now, I see the parallel to actually using "with gil" or "with nogil" inside a function. I was just looking for something that was more explicit about the fact that you wanted functions declared this way to explicitly acquire or release the GIL before being run.

Question: If you declare a function as "with nogil" and the GIL is not held at the time the function is called, does it leave the GIL unlocked after the function returns, preserving its previous state?

**Comment by scoder** [comment]
> If you declare a function as "with nogil" and the GIL is not held at the time the function is called, does it leave the GIL unlocked after the function returns, preserving its previous state?

It definitely should. I think we should start collecting test cases somewhere, BTW.

**Comment by pablogsal** [comment]
@scoder Apologies for the delay, but it has been difficult to allocate time to advance this.

I have started implementing all the ideas discussed here and I have encountered some interesting things to consider:

* We need to decide how "with nogil" interacts with "nogil" functions. This is because "nogil" functions can be called with or without the gil and if they contain a "with nogil" function call, this can crash when called without the gil with a `PyThreadState_Get: no current thread` or similar. As "with nogil" functions basically ensure that the GIL is released when called, we could generate something that checks first the GIL status and then proceeds:

```c
  #ifdef WITH_THREAD
  PyThreadState *_save;
  int __gil_was_held = PyGILState_Check();
  if (__gil_was_held) {
        Py_UNBLOCK_THREADS
        __Pyx_FastGIL_Remember();
  }
  #endif
  #ifdef WITH_THREAD
  if (__gil_was_held) {
        __Pyx_FastGIL_Forget();
        Py_BLOCK_THREADS
  }
  #endif
  __pyx_v_y = sin(__pyx_v_y);
```

* I have tried to implement the feature by trying to inject a `GILStatNode` somewhere in the code for `SimpleCallNode` or by some parse tree transformation but unfortunately it does not glue nicely. This is because `GILStatNode` are expected to enclose statements (they reuse the `TryFinallyStatNode` functionality) and they don't fit nicely with calls. For example, if you surround the `SimpleCallNode` in a `GILStatNode`, then it will crash if the call result is assigned (`x = call()`) because `GILStatNode` don't have the required functionality expected by the assignment node over the `SimpleCallNode`. Additionally, the try-finally node requires the enclosed node to implement `generate_execution_code()` but `SimpleCallNode` only implements `generate_evaluation_code`. I could, of course, hack around this and defer transparently all the code from the `GILStatNode` to its body, but this seems the wrong path. @scoder What are your thoughts here? Is there a preferred path you would like to take here? I can see many ways to "overcome this" but this but I don't have the context of the whole project so I prefer to discuss it first.

**Comment by oliverhaas** [comment]
I stumbled on this issue because I was looking for something like "with nogil functions" when working with fused types. Hopefully I'm confusing people here, but I think there might be a useful addition to the idea suggested here.

Starting with the example from https://cython.readthedocs.io/en/latest/src/userguide/fusedtypes.html#conditional-gil-acquiring-releasing.
```
cimport cython

ctypedef fused double_or_object:
    cython.double
    object

def increment(double_or_object x):
    with nogil(double_or_object is cython.double):
        # Same code handles both cython.double (GIL is released)
        # and python object (GIL is not released).
        x = x + 1
    return x
```
It might be useful to have a "nogil function" for C data types, and just a "normal/gil function" for objects, something like
```
cdef double_or_object increment(double_or_object x) with nogil(double_or_object is cython.double):
    x = x + 1
    return x
```
This should be doable during compile time, but actually implementing this is beyond my skill level. Maybe I'm wrong or overlooking something, but I for sure would use this regularly.
